torch/csrc/jit/tensorexpr/bounds_overlap.cpp

```diff
@@ -6,6 +6,41 @@ namespace jit {
 namespace tensorexpr {
 namespace analysis {
 
+// Returns true if the given expression includes a "Sub" term.
+bool hasSub(ExprPtr e) {
+  return !NodeFinder<Sub>::find(e).empty();
+}
+
+// Returns true if the given expression has negative immediates.
+bool hasNegativeImms(ExprPtr e) {
+#define NEG_IMMS(Type, Name)                    \
+  {                                             \
+    auto imms = NodeFinder<Name##Imm>::find(e); \
+    for (const auto& imm : imms) {              \
+      if (immediateIsNegative(imm)) {           \
+        return true;                            \
+      }                                         \
+    }                                           \
+  }
+  AT_FORALL_SCALAR_TYPES_AND2(Half, BFloat16, NEG_IMMS);
+#undef NEG_IMMS
+  return false;
+}
+
+// Assumption:
+//   * The variables in the given expression can never be negative.
+//
+// So, the only way for the given expression to be negative is if it includes
+// at least one of:
+//     * "subtract" term.
+//     * a negative immediate value.
+// The expression could involve multiple subtract terms and negative
+// immediate values and end up being positive. But even if one of them is
+// present we conservatively assume that it can be negative.
+bool isPositiveExpr(ExprPtr e) {
```

Comment:
Naming nit: `isPositive` sounds like the function is accurate with both `true` and `false`, while in fact we only guarantee the `true`. I.e. `!isPositive` can not be used as `isNegative`. Maybe  `mustBePositive`, `alwaysPositive` or something along these lines?

Reply:
Sounds good. Changed to `mustBePositive`.